routing/chainview/btcd.go

```diff
@@ -241,87 +273,115 @@ func (b *BtcdFilteredChainView) chainFilterer() {
+		}
+		txs := make([]*wire.MsgTx, 0, len(block.Transactions))
+		for _, str := range block.Transactions {
+			b, err := hex.DecodeString(str)
 			if err != nil {
-				log.Errorf("Unable to get block: %v", err)
-				continue
+				return nil, err
 			}
-			b.bestHash, b.bestHeight = *update.blockHash, update.blockHeight
-
-			// Next, we'll scan this block to see if it modified
-			// any of the UTXO set that we're watching.
-			filteredTxns := filterBlock(newBlock)
-
-			// Finally, launch a goroutine to dispatch this
-			// filtered block notification.
-			go func() {
-				b.newBlocks <- &FilteredBlock{
-					Hash:         *update.blockHash,
-					Height:       uint32(update.blockHeight),
-					Transactions: filteredTxns,
-				}
-			}()
+			tx := &wire.MsgTx{}
+			err = tx.Deserialize(bytes.NewReader(b))
+			if err != nil {
+				return nil, err
+			}
+			txs = append(txs, tx)
+		}
+		return &FilteredBlock{
+			Hash:         *hash,
+			Height:       height,
+			Transactions: txs,
+		}, nil
+	}
 
+	for {
+		select {
 		// The caller has just sent an update to the current chain
 		// filter, so we'll apply the update, possibly rewinding our
 		// state partially.
 		case update := <-b.filterUpdates:
+
 			// First, we'll add all the new UTXO's to the set of
 			// watched UTXO's, eliminating any duplicates in the
 			// process.
 			log.Debugf("Updating chain filter with new UTXO's: %v",
 				update.newUtxos)
 			for _, newOp := range update.newUtxos {
+				b.filterMtx.Lock()
 				b.chainFilter[newOp] = struct{}{}
+				b.filterMtx.Unlock()
 			}
 
+			// Apply the new TX filter to btcd, which will cause
+			// all following notifications from and calls to it
+			// return blocks filtered with the new filter.
+			b.btcdConn.LoadTxFilter(false, []btcutil.Address{},
+				update.newUtxos)
+
+			// All blocks gotten after we loaded the filter will
+			// have the filter applied, but we will need to rescan
+			// the blocks up to the height of the block we last
+			// added to the blockQueue.
+			b.bestHeightMtx.Lock()
+			bestHeight := b.bestHeight
+			b.bestHeightMtx.Unlock()
+
 			// If the update height matches our best known height,
 			// then we don't need to do any rewinding.
-			if update.updateHeight == uint32(b.bestHeight) {
+			if update.updateHeight == bestHeight {
 				continue
 			}
 
 			// Otherwise, we'll rewind the state to ensure the
 			// caller doesn't miss any relevant notifications.
 			// Starting from the height _after_ the update height,
-			// we'll walk forwards, manually filtering blocks.
-			for i := int32(update.updateHeight) + 1; i < b.bestHeight+1; i++ {
+			// we'll walk forwards, rescanning one block at a time
+			// with btcd applying the newly loaded filter to each
+			// block.
+			for i := update.updateHeight + 1; i < bestHeight+1; i++ {
 				blockHash, err := b.btcdConn.GetBlockHash(int64(i))
 				if err != nil {
-					log.Errorf("Unable to get block hash: %v", err)
+					log.Errorf("Unable to get block "+
```

Comment:
It might be nice to provide the user with the height of the block in the error message here. Also, because this is expected in certain reorg conditions, it might be better as a `Warnf` instead of `Errorf` (maybe @Roasbeef can chime in on this).

Reply:
Agreed that it would be beneficial to also log the height here!